python/vesin-torch/vesin/torch/metatensor/_neighbors.py

```diff
@@ -0,0 +1,139 @@
+
+    >>> nl.get_neighbor_list(nl_options)
+    TensorBlock
+        samples (54): ['first_atom', 'second_atom', 'cell_shift_a', 'cell_shift_b',
+    ...    'cell_shift_c']
+        components (3): ['xyz']
+        properties (1): ['distance']
+        gradients: None
+    """
+
+    def __init__(self, nl_options: NeighborListOptions):
+        if not torch.jit.is_scripting():
+            if not _has_mta:
+                raise ModuleNotFoundError(
+                    "`vesin.metatensor` requires the `metatensor-torch` package"
+                )
+        self._nl_options = nl_options
+        self._nl = NeighborListTorch(
+            cutoff=nl_options.cutoff, full_list=nl_options.full_list
+        )
+
+    def compute(
+        self,
+        system: System,
+    ) -> TensorBlock:
+        """
+        Compute the neighbor list for the :py:class:`System
+        <metatensor.torch.atomistic.System>` ``system`` and adds
+        to its neighbor lists. If the system arrays require gradients, the
+        gradients to the distance vectors will be also added to the
+        computational graph.
+
+        :param system: a :py:class:`System
+            <metatensor.torch.atomistic.System>` object containing the
+            data about a structure.
+
+        :return: a :py:class:`metatensor.TensorBlock` containing
+            the neighbor list data. The data will also be added to
+            the set of neighbor lists associated with ``system``.
+        """
+
+        # move to float64, as vesin only works in torch64
+        points = system.positions.to(torch.float64)
+        box = system.cell.to(torch.float64)
+        if torch.all(system.pbc):
+            periodic = True
+        elif not torch.any(system.pbc):
+            periodic = False
+        else:
+            raise NotImplementedError("`vesin` currently does not support mixed pbcs.")
+
+        # computes neighbor list
+        (
+            i,
+            j,
+            S,
+            D,
+        ) = self._nl.compute(
+            points=points,
+            box=box,
+            periodic=periodic,
+            quantities="ijSD",
+            copy=True,
+        )
+
+        # converts to a suitable TensorBlock format
+        nl_block = TensorBlock(
+            D.reshape(-1, 3, 1).to(system.positions.dtype),
+            samples=Labels(
+                names=[
+                    "first_atom",
+                    "second_atom",
+                    "cell_shift_a",
+                    "cell_shift_b",
+                    "cell_shift_c",
+                ],
+                values=torch.hstack([i.reshape(-1, 1), j.reshape(-1, 1), S]),
+            ),
+            components=[Labels.range("xyz", 3)],
+            properties=Labels(names=["distance"], values=torch.tensor([[0]])),
+        )
+
+        # store into the System and return
+        if system.positions.requires_grad or system.cell.requires_grad:
+            register_autograd_neighbors(system, nl_block)
```

Comment:
I don't think we need this one, since this repo contains the same hack already and uses it by default for the `vesin.torch` calculator. Worth adding a small gradcheck test though. Maybe just trying to compute the gradient of NL w.r.t. positions, which will fail if we are missing a part of the graph?

Reply:
fixed and added test